### PR TITLE
Move down breakpoints when placed on empty/skipped lines

### PR DESCRIPTION
**Pull request by haltman-at**
This PR addresses issue #1317, at least partially.  It introduces a selector, `controller.breakpoints.resolver`, that, given a breakpoint, returns where it should be moved down to.  The CLI now runs all breakpoints through `controller.breakpoints.resolver` before adding them (and reports back to the user where the breakpoint was *actually* added).  Note that the resolver may return `null` if there is nowhere suitable below to place a breakpoint; in this case the CLI will not add a breakpoint.

Note that the resolver does not work perfectly -- it's something of a hack -- but it should be good enough for common cases; see below for more explanation.  Also note that attempts to *remove* breakpoints are not run through the resolver.  But that's OK, because (in addition to the messages above), **also** added in this PR is (finally) a way to view all breakpoints!  Now, typing `?` will print not only all watch expressions, but also all breakpoints.

(For node-based breakpoints, the location will simply be reported as "a point in line [number]" (or "a point in line [number] in [file]" if it's in a different file), unless you're currently standing on the breakpoint, in which case it will be reported as "*this* point in line n" (emphasis added).)

One more change in this PR: I decided that the selector `controller.finished` should probably really be `controller.current.finished`.  Of course, if this is considered too much of a breaking change, that'll be an easy change to revert!  Let me know if I should.

So, how's the resolver work?

Well, for a node-based breakpoint, it simply returns the breakpoint that was entered.  (Yes, even if you're currently standing on a skipped node, due to use of `;`; perhaps such breakpoints should also be moved down, but I'm assuming that that case just isn't going to happen that often.)  For line-based breakpoints, it checks whether there any nodes starting on that line that it judges to be non-skipped.  If not, it moves it down a line and repeats, until it either finds a place to put it or runs out of lines.  (Note we go by what line the node *starts* on because that's what `continueUntilBreakpoint` goes by.)

This leaves two questions: How does it determine what nodes start on that line, and how does it determine whether a node is skipped or not?

For the first question, I altered `ast/map.js` a bit.  I took the bit where it finds which nodes overlap the range and factored that out into its own function, `findOverlappingRange`, which `findRange` now calls.  I then added `anyNonSkippedInRange`, which calls `findOverlappingRange` (here the range passed in is a full line), and checks if any of the results *start* in that range and are non-skipped.

Here's the hacky part -- what counts as a skipped node?  We can't really determine this with perfect reliability; for instance, references to constant state variables are often skipped, but I didn't think it was worth it to include that case here (especially as that's unlikely to be a full line); and certain type conversions are skipped, but it was *definitely* not worth it to include that (determining whether we're dealing with a skipped or non-skipped type conversion would just be a pain).

And what about, say, pragmas?  Surely one shouldn't be able to place a breakpoint on a line with only a pragma!  But on the other hand, probably nobody's going to attempt to do that.  So I didn't bother marking those as skipped.  One could go on in that vein -- event declarations? custom type declarations? -- but again, not worth it.

So what *did* I mark as skipped?  I decided to mark as skipped:
1. The node types we deliberately skip over: `ContractDefinition` and `VariableDeclaration`.
2. Anything else that was necessary to make skipping over variable declarations work, which meant `VariableDeclarationStatement`, and the various `[...]TypeName` nodes (and `Mapping`, because a mapping type name node is called `Mapping`, not `MappingTypeName`, for whatever reason).
3. String literals, because those are kind of infamous for their skippedness.

Note by the way that the list of deliberately skipped nodes has been moved out of `controller/sagas` and into `helpers` since that's no longer the only place that uses it (it's also been replaced with a function).

Like I said, it's not perfect, but I think it should be good enough for the common cases that originally gave rise to #1317.

(I also added a test.)

### CODE REVIEW

**Comment by gnidan** [comment]
> I decided that the selector controller.finished should probably really be controller.current.finished

We could always make `controller.finished` an alias. At some point it might be worthwhile to add some kind of way to mark selectors as deprecated in reselect-tree.

File: packages/truffle-core/lib/commands/debug.js [hunk 1]
```diff
@@ -730,6 +771,7 @@ var command = {
                 break;
               case "?":
                 printWatchExpressions();
+                printBreakpoints();
```
Comment (gnidan) [inline packages/truffle-core/lib/commands/debug.js]: For future PRs: it's helpful to upload a screenshot of output changes so reviewers don't have to grab the branch locally just to see relevant info. Looks good here though.

**Comment by haltman-at** [comment]
> > I decided that the selector controller.finished should probably really be controller.current.finished
> 
> We could always make `controller.finished` an alias. At some point it might be worthwhile to add some kind of way to mark selectors as deprecated in reselect-tree.

Good point, I'll add that back in.

File: packages/truffle-core/lib/commands/debug.js (hunk 1 above)
Comment (haltman-at) [reply packages/truffle-core/lib/commands/debug.js]: Oh, I see!  Didn't realize I could do that.  Noted.

**Pull request merged**